staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go

```diff
@@ -35,72 +36,121 @@ func init() {
+	options, err := openstack.AuthOptionsFromEnv()
+	if err != nil {
+		return nil, fmt.Errorf("error reading openstack env vars: %s", err)
+	}
+	tg := &tokenGetter{options}
+	ctg := newCachedGetter(tg, persister, oapConfig)
+	return &openstackAuthProvider{
+		tokenGetter: ctg,
+		persister:   persister,
+	}, nil
 }
 
+// Login is not used
+func (oap *openstackAuthProvider) Login() error { return nil }
+
 // TokenGetter returns a bearer token that can be inserted into request.
 type TokenGetter interface {
-	Token() (string, error)
+	Token() (*openstackToken, error)
 }
 
-type tokenGetter struct{}
+type tokenGetter struct {
+	options gophercloud.AuthOptions
+}
 
-// Token creates a token by authenticate with keystone.
-func (*tokenGetter) Token() (string, error) {
-	options, err := openstack.AuthOptionsFromEnv()
+// Token gets a token from the OpenStack Identity service
+func (tg *tokenGetter) Token() (*openstackToken, error) {
+	token, err := tokenFromAuth(tg.options)
 	if err != nil {
-		return "", fmt.Errorf("failed to read openstack env vars: %s", err)
+		return &openstackToken{}, fmt.Errorf("error getting openstack token: %s", err)
 	}
-	client, err := openstack.AuthenticatedClient(options)
-	if err != nil {
-		return "", fmt.Errorf("authentication failed: %s", err)
-	}
-	return client.TokenID, nil
+	return token, nil
 }
 
-// cachedGetter caches a token until it gets expired, after the expiration, it will
-// generate another token and cache it.
 type cachedGetter struct {
-	mutex       sync.Mutex
-	tokenGetter TokenGetter
+	lk        sync.Mutex
+	tg        TokenGetter
+	persister restclient.AuthProviderConfigPersister
+	cache     map[string]string
+}
 
-	token string
-	born  time.Time
-	ttl   time.Duration
+func newCachedGetter(tg TokenGetter, persister restclient.AuthProviderConfigPersister, cache map[string]string) *cachedGetter {
+	if cache == nil {
+		cache = make(map[string]string)
+	}
+	return &cachedGetter{
+		tg:        tg,
+		persister: persister,
+		cache:     cache,
+	}
 }
 
 // Token returns the current available token, create a new one if expired.
-func (c *cachedGetter) Token() (string, error) {
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-
-	var err error
-	// no token or exceeds the TTL
-	if c.token == "" || time.Now().Sub(c.born) > c.ttl {
-		c.token, err = c.tokenGetter.Token()
-		if err != nil {
-			return "", fmt.Errorf("failed to get token: %s", err)
+func (g *cachedGetter) Token() (*openstackToken, error) {
+	token := g.cachedToken()
+	if token.Valid() {
+		return token, nil
+	}
+	token, err := g.tg.Token()
+	if err != nil {
+		return nil, err
+	}
+	cache := g.update(token)
+	if g.persister != nil {
+		if err := g.persister.Persist(cache); err != nil {
+			glog.V(4).Infof("Failed to persist token: %v", err)
 		}
-		c.born = time.Now()
 	}
-	return c.token, nil
+	return token, nil
+}
+
+func (g *cachedGetter) cachedToken() *openstackToken {
```

Comment:
Good point. Having `cachedToken()` return an error cleans up the method itself.

But I think the `Valid` method is correct as-is, since a nil token or one whose ID is an empty string is indeed invalid.